src/Swarm/Game/Scenario/Objective/Presentation/Render.hs

```diff
@@ -0,0 +1,87 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- Display logic for Objectives.
+module Swarm.Game.Scenario.Objective.Presentation.Render where
+
+import Brick hiding (Direction, Location)
+import Brick.Widgets.Center
+import Brick.Widgets.List qualified as BL
+import Control.Lens hiding (Const, from)
+import Data.List.NonEmpty qualified as NE
+import Data.Map.Strict qualified as M
+import Data.Maybe (listToMaybe)
+import Data.Text (Text)
+import Data.Text qualified as T
+import Data.Vector qualified as V
+import Swarm.Game.Scenario.Objective
+import Swarm.Game.Scenario.Objective.Presentation.Model
+import Swarm.TUI.Attr
+import Swarm.TUI.Model.Name
+import Swarm.TUI.View.Util
+
+makeListWidget :: GoalTracking -> BL.List Name GoalEntry
+makeListWidget (GoalTracking _announcements categorizedObjs) =
+  BL.listMoveTo 1 $ BL.list ObjectivesList (V.fromList objList) 1
+ where
+  objList = concatMap f $ M.toList categorizedObjs
+  f (h, xs) = Header h : map (Goal h) (NE.toList xs)
+
+renderGoalsDisplay :: GoalDisplay -> Widget Name
+renderGoalsDisplay gd =
+  padAll 1 $
+    hBox
+      [ hLimitPercent 30 $
+          vBox
+            [ hCenter $ str "Goals"
+            , padAll 1 $
+                vLimit 10 $
+                  BL.renderList (const drawGoalListItem) True lw
+            ]
+      , hLimitPercent 70 $
+          padLeft (Pad 2) $
+            maybe emptyWidget (singleGoalDetails . snd) $
+              BL.listSelectedElement lw
+      ]
+ where
+  lw = _listWidget gd
+
+withEllipsis :: Text -> Widget Name
```

Comment:
This seems like a nice utility we should factor out into e.g. `Swarm.TUI.View.Util`.